Debezium Community Stories With... Renato Mefi (#503)
debezium/debezium.github.io: blog/2020-10-08-debezium-community-stories-with-renato-mefi.adoc

On the internet you're going to find me talking about Docker, Debezium, Kafka, Microservices and other things that I enjoy.
Although those amazing engineering pieces really excite me, at this moment I'm also really passionate about Platform Engineering teams and how they can operate in an organization, the stories I'm going to tell below represent my view of it, of how critical the role of a platform team can be when adopting new technologies and solving difficult problems for the whole, in this case powered by Debezium!

*What are your use cases for Debezium and CDC in your current project?*

It's a long and enjoyable story (long in terms of the internet), we've been using Debezium since Q4 2018, it's been 2 years at the moment I'm writing those answers here.

When I classify Debezium within our product, I say it is an architectural component, the idea behind this is to position it as a platform/infrastructure concern, in a way that it can reach multiple parts of the stack and services. I consider this abstraction of Debezium one of the key success factors it had in its adoption and growth within our platform, let me explain this better!

> Our first use case is likely to be one of the most common ones for CDC, the https://martinfowler.com/bliki/StranglerFigApplication.html[strangler pattern], which for us came before Debezium; so let me tell this part of the story first: when I joined Usabilla (later acquired by SurveyMoney), there was already an effort to move our platform to a new architecture and the strangler pattern was already there, when the first couple of services started to grow, their primary way to bring data out of legacy was to poll the database, and needless to say, this could go very wrong! Our legacy database is a MongoDB cluster, and since I was preoccupied with the polling approach I started to dig into possibilities, I was hoping to find something like a streaming api for it, but what I ended up encountering was the database changelog, "oplog" as it's called in Mongo!

Reviewer: ```suggestion
Our first use case is likely to be one of the most common ones for CDC, the https://martinfowler.com/bliki/StranglerFigApplication.html[strangler pattern], which for us came before Debezium; so let me tell this part of the story first: when I joined Usabilla (later acquired by SurveyMoney), there was already an effort to move our platform to a new architecture and the strangler pattern was already there, when the first couple of services started to grow, their primary way to bring data out of legacy was to poll the database, and needless to say, this could go very wrong! Our legacy database is a MongoDB cluster, and since I was preoccupied with the polling approach I started to dig into possibilities, I was hoping to find something like a streaming api for it, but what I ended up encountering was the database changelog (link:https://en.wikipedia.org/wiki/Write-ahead_logging[Write-ahead logging], "oplog" as it's called in Mongo!
```

Now back to our use case, what makes it an architecture component for us, is basically the approach, we abstracted and wrapped Debezium in a project called Legacy Data Syncer (LDS for us, because acronyms never get old). Although it might look simple to spin up a Kafka Connect with Debezium, running it production ready, monitoring multiple collections within the database, exposing metrics, doing transformations and more, is not such an easy task. So how does it work? Every time an engineering team needs to capture data from our legacy system, to start strangling a feature, they only have to do two things, open a pull request which literally adds one line to LDS, and create their Kafka consumer! 

[.centered-image.responsive-image]
====
++++

> <img src="/images/debezium_community_stories_with_renato_mefi_lds.png" style="max-width:100%;" class="responsive-image">

Author: @renatomefi, could you perhaps provide that image with a larger font size/resolution? It's rather small right now.

Reviewer: I'm going to add this image to carbon.now.sh and the outbox one as well!
This way we can even match the colors to debezium style :)

Reviewer: Legacy data syncer: https://carbon.now.sh/?bg=rgba(171%2C%20184%2C%20195%2C%201)&t=lucario&wt=none&l=yaml&ds=true&dsyoff=20px&dsblur=68px&wc=true&wa=true&pv=56px&ph=56px&ln=false&fl=1&fm=Hack&fs=14px&lh=133%25&si=false&es=2x&wm=false&code=%250Aconnectors%253A%250A%2520%2520config%253A%250A%2520%2520%2509%2523%2520Add%2520a%2520new%2520line%2520to%2520monitor%2520a%2520new%2520collection%2520in%2520MongoDB%250A%2520%2520%2520%2520%2523%2520Template%253A%250A%2520%2520%2520%2520%2523%2520%2522Collection%2520name%2522%253A%2520%255B%2522Enable%2520raw%2520Oplog%2522%252C%2520%2522Enable%2520CDC%2520transformation%2522%255D%250A%2520%2520%2520%2520%250A%2520%2520%2520%2520collection_a%253A%2520%255B%2522op%2522%252C%2520%2522cdc%2522%255D%250A%2520%2520%2520%2520%2523%2520derived%2520topics%253A%2520%2560legacy-data-syncer.op.collection-a%2560%2520%2560legacy-data-syncer.cdc.collection-a%2560%250A%2520%2520%2520%2520%2523%2520connector%2520names%253A%2520%2560op-collection_a%2560%2520%2560cdc-collection_a%2560%250A%2520%2520%2520%2520%250A%2520%2520%2520%2520collection_b%253A%2520%255B%2522op%2522%252C%2520%2522cdc%2522%255D%250A%2520%2520%2520%2520collection_c%253A%2520%255B%2522op%2522%252C%2520%2522cdc%2522%255D

++++
*Figure 1. The configuration file in LDS; a developer will open a PR adding a new line, the rest will be taken care of.*
====

> Upon merging the PR, our project will provision the whole configuration to Kafka Connect, ensures the snapshot is executed, metrics and presents and etc; We've done the same thing for the outbox pattern and I talk a little bit more about it in this https://twitter.com/renatomefi/status/1185098904745992197[tweet thread].

Reviewer: ```suggestion
Upon merging the PR, our project will provision the whole configuration to Kafka Connect, ensures the snapshot is executed, metrics are present and etc; We've done the same thing for the outbox pattern and I talk a little bit more about it in this https://twitter.com/renatomefi/status/1185098904745992197[tweet thread].
```

*You mention the outbox pattern; Could you tell more about why and how you're using this?*

Absolutely! One more time, it's crazy how CDC and Debezium can simplify some of the most critical architectural parts of big platforms!
One year after using Debezium in the core of our architecture migration, we had another problem at our hands: how to reliably write data to our new source of truth databases and propagate messages to Kafka at the same time. Although it seems to be simple to answer and find a solution, each of them comes with a major drawback.

Which solutions do we have?

* Embrace eventual consistency to its peak by adopting _Event Sourcing_, by writing first to Kafka and reading our own writes, the drawbacks here are extra complexity and intensified eventual consistency
* _Dual writes_, well, actually this is not an option, because as you know, https://thorben-janssen.com/dual-writes/["Friends Don't Let Friends Do Dual Writes"]!

> * Different approaches of _distributed transactions_ like 2PC and sagas; the costs here are performance and engineering effort, now every service I have has to either become a transaction coordinator or have rollback capabilities, also the cascade effect scares us quite a bit!

Reviewer: ```suggestion
* Different approaches of _distributed transactions_ like 2PC and sagas; the costs here are performance and engineering effort, now every service we have has to either become a transaction coordinator or have rollback capabilities, also the cascade effect scares us quite a bit!
```

Well, what's left? We found that outbox was the right answer for us, but before we get there, let me get into the cost x benefit equation of our decision making!

Although some of the options were quite attractive technically, for instance event sourcing, the engineering effort and growth is immense. Also, it's not the kind of thing which comes ready to use, and there's a lot of discovery to be made along the way, so what were the constraints and desires:

* _Reliability_; we want at least once semantics, exactly once isn't necessary as we can uniquely identify each message/event;

> * _Eventual consistency only between services_, but not within the services themselves. Being able to interact with a service which is the source of truth of a certain model, and get an immediate answer is not just handy, but incredibly powerful;

Reviewer: ```suggestion
* _Eventual consistency only between services_, but not within the services themselves. Being able to interact with a service which is the source of truth of a certain model, and get an immediate answer is not just handy, but incredibly powerful (and that's why monoliths are also so attractive);
```

We realized that the outbox pattern would help us meet those requirements: applications would publish events via an outbox table, which gets written to as part of business transactions in the database.

As with the strangler pattern, we wanted to resort to an architecture component, something the teams could self-service. At first, we were exploring a home-grown solution which would look for the outbox tables among every service and publish the messages. The problem with this approach would be the polling databases problem, although in this case this is less harmful as we don't need to look for updates or deletes.

Luckily, by that time I was closely following the work being done in Debezium and I read the blogpost about link:/blog/2019/02/19/reliable-microservices-data-exchange-with-the-outbox-pattern/[reliable data exchange between microservices using the outbox pattern], and there was my answer! Well, I mean, parts of the answer, we still needed to implement it, and that's a story for the next question!

> Fast forward a couple of months and we got a reliable way to exchange messages between services, with all the guarantees we wanted to have, and by applying some platform DeVops flavor to it, we also made it self-service and easy to plug in every service!

Author: "we also made it self-service and easy to plug in every service" -- would you happen to have a screenshot of that, too?

> Fast forward a couple of months and we got a reliable way to exchange messages between services, with all the guarantees we wanted to have, and by applying some platform DeVops flavor to it, we also made it self-service and easy to plug in every service!

Reviewer: ```suggestion
Fast forward a couple of months and we got a reliable way to exchange messages between services, with all the guarantees we wanted to have, and by applying some platform DevOps flavor to it, we also made it self-service and easy to plug in every service!

The user can specify which database their service is at, what's the table name and which column to use as even router, you can find more details about it in the official link:https://debezium.io/documentation/reference/configuration/outbox-event-router.html#outbox-event-router-property-route-by-field[Debezium Outbox Event Router docs].
```

Reviewer: And the example configuration: https://carbon.now.sh/?bg=rgba(171%2C%20184%2C%20195%2C%201)&t=lucario&wt=none&l=yaml&ds=true&dsyoff=20px&dsblur=68px&wc=true&wa=true&pv=56px&ph=56px&ln=false&fl=1&fm=Hack&fs=14px&lh=133%25&si=false&es=2x&wm=false&code=%250Aconnectors%253A%250A%2520%2520config%253A%250A%2520%2520%2520%2520%2523%2520Add%2520a%2520new%2520object%2520to%2520monitor%2520the%2520Outbox%2520table%2520from%2520a%2520given%2520Service%250A%2520%2520%2509test-connector%253A%2520%2523%2520Unique%252C%2520usually%2520the%2520microservice%2520name%250A%2520%2520%2509%2509database%253A%2520my_service_database_name%250A%2520%2520%2520%2520%2520%2520%2520%2520schema%253A%2520my_service_schema_name%250A%2520%2520%2520%2520%2520%2520%2520%2520table%253A%2520outbox%250A%2520%2520%2520%2520%2520%2520%2520%2520route_by_field%253A%2520message_route

*You're not only using Debezium but you've also contributed to the project. How was your experience doing so? Are you doing other open-source work, too?*

As I spoiled at the beginning, my usage and contributions to Debezium walked hand-to-hand, in both the use cases we have for Debezium in SurveyMonkey I had great opportunities to contribute to both Debezium and Kafka (just a bug fix, but I'm happy about it!).

> At first I was fixing bugs in the Debezium MongoDB connector; as we really scaled it up to all the teams a lot of edge cases started to show up, mostly in the transform which takes the raw database transaction log and transforms into a nicely readable Kafka Connect struct. Also due to our architecture choice, we split the raw log and transformed data into two different steps, which go in separate topics and are configured as separate Kafka Connect connectors. That also created some extra tiny features and challenges to be incorporated in Debezium itself!

Author: "separate Kafka Connect connectors" -- could you still clarify this part; you mentioned something around Mirus?

> At first I was fixing bugs in the Debezium MongoDB connector; as we really scaled it up to all the teams a lot of edge cases started to show up, mostly in the transform which takes the raw database transaction log and transforms into a nicely readable Kafka Connect struct. Also due to our architecture choice, we split the raw log and transformed data into two different steps, which go in separate topics and are configured as separate Kafka Connect connectors. That also created some extra tiny features and challenges to be incorporated in Debezium itself!

Reviewer: ```suggestion
At first I was fixing bugs in the Debezium MongoDB connector; as we really scaled it up to all the teams a lot of edge cases started to show up, mostly in the transformation which takes the raw database transaction log and transforms into a nicely readable Kafka Connect struct. Also due to our architecture choice, we split the raw log and transformed data into two different steps, which go in separate topics and are configured as separate Kafka Connect connectors.

The step we call `op` (stands for operation), is the Debezium MongoDB sink connector and outputs the raw data into the topic without any change or transformation, minimizing the chances of errors in the process, the second step called `cdc`, is a link:https://github.com/salesforce/mirus[Salesforce Mirus] source connector, which reads from the `op` output topic, transforms the message using the link:https://debezium.io/documentation/reference/1.3/configuration/mongodb-event-flattening.html[Debezium Document SMT] and outputs to the final topic, which the services can consume from. With this approach, we have two main abilities: Resist to errors and crashes on the native/custom transformation process, and have the chance to change the transformation to our desires without having to read from the database again, given us more flexibility. That also created some extra features and challenges to be incorporated in Debezium itself!
```

> Quick sidestep: the rationale behind this decision was to be able to survive transformation errors; MongoDB has a replication window which, if you lose it, means that you are going to have to make a new full snapshot of the collection and you might lose deletion events in this process. Because of this we opted for a safer approach, which was to split the logic of transformation from the raw logs.

Reviewer: I realized this paragraph mixes with the new suggestion above, let me see how to merge them

> At first I was fixing bugs in the Debezium MongoDB connector; as we really scaled it up to all the teams a lot of edge cases started to show up, mostly in the transform which takes the raw database transaction log and transforms into a nicely readable Kafka Connect struct. Also due to our architecture choice, we split the raw log and transformed data into two different steps, which go in separate topics and are configured as separate Kafka Connect connectors. That also created some extra tiny features and challenges to be incorporated in Debezium itself!
> 
> Quick sidestep: the rationale behind this decision was to be able to survive transformation errors; MongoDB has a replication window which, if you lose it, means that you are going to have to make a new full snapshot of the collection and you might lose deletion events in this process. Because of this we opted for a safer approach, which was to split the logic of transformation from the raw logs.

Reviewer: ```suggestion
At first, I was fixing bugs in the Debezium MongoDB connector; as we really scaled it up to all the teams a lot of edge cases started to show up, mostly in the transformation which takes the raw database transaction log and transforms into a nicely readable Kafka Connect struct. Also due to our architecture choice, we split the raw log and transformed data into two different steps, which go in separate topics and are configured as separate Kafka Connect connectors.

Quick sidestep: the rationale behind this decision was to be able to survive transformation errors; MongoDB has a replication window which, if you lose it, means that you are going to have to make a new full snapshot of the collection and you might lose deletion events in this process. Because of this we opted for a safer approach, which was to split the logic of transformation from the raw logs like this:
The step we call `op` (stands for operation), is the Debezium MongoDB sink connector and outputs the raw data into the topic without any change or transformation, minimizing the chances of errors in the process, the second step called `cdc`, is a link:https://github.com/salesforce/mirus[Salesforce Mirus] source connector, which reads from the `op` output topic, transforms the message using the link:https://debezium.io/documentation/reference/1.3/configuration/mongodb-event-flattening.html[Debezium Document SMT] and outputs to the final topic, which the services can consume from. With this approach, we now have two main abilities: Resist to errors and crashes on the native/custom transformation process like mentioned above, and have the chance to change the transformation to our desires without having to read from the database again, given us more flexibility. That also created some extra features and challenges to be incorporated in Debezium itself!
```

> As I kept contributing I noticed a few things that could be improved and started fixing them, including an almost full refactor of the build of Debezium's container  images, its scripts and other smaller things!

Reviewer: ```suggestion
As I kept contributing I noticed a few things that could be improved and started fixing them, including an almost full refactor of the build of Debezium's container images, its scripts, and other smaller things!
```

Will I do more open source? Yes, but I must say most of my open source activity is "selfish", I'm developing solutions to problems I face at work but I'm happy to take the extra step and make them to the OSS world, but it also makes it seasonal. One of the advantages to that is if I'm doing something for a project, be sure I'll make it to production and likely be able to find more corner cases!

*Is there anything you're missing in Debezium or you'd like to see improved in the future?*

When I think of the Kafka and Debezium ecosystem, the next steps I consider important are the ones which will make it more accessible. Although there's a lot of content and examples online, there's still a big gap between reading those and getting to a production ready implementation.

What I mean by that is abstracting the individual pieces away and giving them more meaning. The outbox pattern is a good example, it was not natural for people to think of CDC and know that it was such a good match to it, there are plenty of more use cases to be explored in this ecosystem.

> What if you could have everything out-of-the-box? An outbox implementation in your favorite framework, which knows how to integrate with the ORM, handle the transaction part, then, how to shape the messages and events? How to adopt the schema for it and how an evolution of it looks like. After that, getting closer to the consumer implementation, how can I handle the messages idempotently, respect the semantics, do retries and project them to a database if need be?

Reviewer: ```suggestion
What if you could have everything out-of-the-box? An outbox implementation in your favorite framework, which knows how to integrate with the ORM, handle the transaction part, then, how to shape the messages and events? How to adopt the schema for it and how an evolution of it looks like. After that, getting closer to the consumer implementation, how can I handle the messages idempotently, respect the semantics, do retries, and project them to a database if need be? There are already initiatives like those, for instance, the link:https://debezium.io/documentation/reference/integrations/outbox.html[Quarkus Outbox Extension], which takes care of framework and database integration. The future for me has those things, for multiple frameworks and tech stacks, going even broader and helping you design good events (maybe even powered by link:https://www.asyncapi.com/ [AsyncAPI]), giving everyone a kickstart!
```
